### ISSUE TITLE
Directory paths are ignored even if negative-ignore exists for files underneath it

### ISSUE THREAD

**Issue by infinity0** [BurntSushi/ripgrep#1050]
Running over https://salsa.debian.org/rust-team/debcargo-conf/:

~~~~
$ cat .gitignore
/build
/src/*/*
!/src/*/debian
$ rg -q copyright 
$ rg -q copyright src
No files were searched, which means ripgrep probably applied a filter you didn't expect. Try running again with --debug.
1
$ rg -q copyright src/*
$ git grep -q copyright 
$ git grep -q copyright src
$ git grep -q copyright src/*
~~~~

**Comment by BurntSushi** [comment]
Yeah, this is one of the cases where it's tough for ripgrep to behave the same as git here. I think there have been bugs reported on this issue in the past, but I can't find them.

The fundamental issue is that `/src/*/*` matches `src/atty` (for example), and therefore, that directory is ignored and ripgrep doesn't descend into it. So the `!/src/*/debian` whitelist rule isn't even applied.

Two thoughts:

1. Should `src/atty` actually match `/src/*/*`? There might be some special handling going on with the trailing slash that allows it to match. Not sure.
2. Maybe ripgrep either needs to get smarter, or it cannot implement the optimization of avoiding to descend into directories that have been ignored.

**Comment by infinity0** [comment]
> Should src/atty actually match /src/*/*?

No, */* is pretty explicit about only matching 2 levels down. I think this would take it further away from the behaviour of git.

> Maybe ripgrep either needs to get smarter, or it cannot implement the optimization of avoiding to descend into directories that have been ignored.

I think this would be the correct approach, how about not implementing the optimisation if there exists a ! whitelist rule matching potentially anything below that directory?

**Comment by BurntSushi** [comment]
> how about not implementing the optimisation if there exists a ! whitelist rule matching potentially anything below that directory?

Yes, that's what I mean by "smarter." There's no infrastructure for that kind of analysis yet.

**Comment by ssbarnea** [comment]
In fact the ignore-file implementation seems to be quite broken, I just discovered that an entry like `.pre-commit/` caused the ignore of all files in all folders named `.pre-commit-config.yaml`, which is clearly not what I was looking for.

**Comment by BurntSushi** [comment]
> In fact the ignore-file implementation seems to be quite broken, I just discovered that an entry like `.pre-commit/` caused the ignore of all files in all folders named `.pre-commit-config.yaml`, which is clearly not what I was looking for.

That is entirely separate from this issue. Please open a new issue with an MRE.

**Comment by sharifhsn** [comment]
Closed by 14f4957

Repro:

```sh
git init -q
mkdir -p build src/atty/debian src/atty/foo
printf '/build\n/src/*/*\n!/src/*/debian\n' > .gitignore
printf 'copyright text\n' > src/atty/debian/copyright
printf 'copyright text\n' > src/atty/foo/bar
```

`src/atty/foo/bar` matches `/src/*/*` and should be ignored, only the whitelisted `src/atty/debian/copyright` should be searched.

ripgrep 14.1.1:

```
$ rg copyright src
src/atty/foo/bar:copyright text
src/atty/debian/copyright:copyright text
```

ripgrep 15.0.0:

```
$ rg copyright src
src/atty/debian/copyright:copyright text
```
